### ISSUE TITLE
[Security] SpanTap.uninstall() never resets `recordRootSpans` / `captureMessagePayloads`, so detaching DevTools leaves message payloads attached to spans a pre-existing third-party tracer keeps exporting

### ISSUE THREAD

**Issue by pathosDev** [pathosDev/actor-ts#714]
**Component:** `src/devtools/taps/SpanTap.ts`
**Severity (assessment):** MEDIUM
**CWE:** CWE-532 / CWE-200 (Exposure of Sensitive Information)

`SpanTap.install()` switches on two system-wide tracing flags. `SpanTap.uninstall()` restores the previous *tracer* but only resets the flags on the branch where it also calls `TracingExtension.disable()` — i.e. only when no tracer was installed beforehand. In the production shape (an OTel adapter already exporting to an APM), `uninstall()` takes the `tracing.enable(this.previousTracer)` branch, which does not touch `_traceRootSpans` / `_traceMessagePayloads`. Both stay `true` for the remaining life of the process.

### Exploit walkthrough

Preconditions: the application enables tracing in production (`TracingExtensionId.enable(otelTracer(...))` — the documented production path, exporting spans to Jaeger/Tempo/Honeycomb/Datadog), and an operator attaches DevTools once (`DevTools.attach(...)`) with the `tracing` panel enabled (on by default), then detaches.

1. Before attach: `_traceMessagePayloads = false`; spans carry only `actor.path`, `actor.message.type`, `actor.sender`.
2. Operator attaches DevTools to debug an incident. `SpanTap.install()` runs, wraps the OTel tracer in a `TeeTracer`, and sets both flags.
3. Operator detaches (`DevTools.attach(...).detach()` → `DevToolsServer.stop()` → `tap.uninstall()`). The OTel adapter is correctly restored — and both flags remain on.
4. From now on every single actor message opens a root span and carries `actor.message.payload`, a JSON serialisation of the whole message (depth 6, truncated at the length limit) — order bodies, auth tokens carried in commands, PII, anything an actor receives — and the restored OTel exporter ships all of it to the external trace vendor. There is no log line, no metric, and no API that reveals it other than calling `isCapturingMessagePayloads()`; only a process restart clears it.
5. Secondary impact: `recordRootSpans` also stays on, so the span volume equals the total message rate rather than the sampled application traces — an unbudgeted cost/ingest blowup at the APM.

Verified against the real `SpanTap` and `TracingExtension` under Bun (probe at scratchpad/probe-spantap-leak.ts):
```
baseline  payloads= false rootSpans= false
attached  payloads= true rootSpans= true
detached  payloads= true rootSpans= true | tracer restored = true
```

### Evidence — `src/devtools/taps/SpanTap.ts`

src/devtools/taps/SpanTap.ts:67-88 (install turns both on unconditionally):
```ts
this.previousTracer = tracing.isEnabled() ? tracing.get() : null;
if (this.previousTracer === null) {
  tracing.enable(new RecordingTracer({ maxRecorded: 0, onSpanEnd: (span) => this.onSpanEnd(span) }));
} else {
  tracing.enable(new TeeTracer(this.previousTracer, (span) => this.onSpanEnd(span)));
}
this.installed = true;
tracing.recordRootSpans(true);
tracing.captureMessagePayloads(true);
```

src/devtools/taps/SpanTap.ts:97-108 (uninstall — the `else` branch never resets the flags):
```ts
uninstall(): void {
  ...
  const tracing = this.system.extension(TracingExtensionId);
  if (this.previousTracer === null) tracing.disable();
  else tracing.enable(this.previousTracer);
  this.previousTracer = null;
  ...
}
```

src/tracing/TracingExtension.ts:24-27 vs 76-82 — only `disable()` clears the flags; `enable()` does not:
```ts
enable(tracer: Tracer): Tracer {
  this.tracer = tracer;
  return tracer;
}
...
disable(): void {
  this.tracer = NOOP_TRACER;
  this._system._traceRootSpans = false;
  this._system._traceMessagePayloads = false;
}
```

src/internal/ActorCell.ts:940-957 — what the flag actually emits on every message:
```ts
if (this.system._traceMessagePayloads) {
  const payload = describeMessagePayload(message);
  if (payload !== null) attributes['actor.message.payload'] = payload;
}
span = tracer.startSpan('actor.receive', { parent: env.trace ?? undefined, kind: 'consumer', attributes });
```

The documentation states the opposite behaviour — docs/src/content/docs/observability/devtools/tracing.mdx: "it opens a root span for every message and serialises each payload, and it does so for as long as DevTools is attached." And docs/src/content/docs/observability/tracing/actor-tracing.mdx carries a caution built on the same (now false) premise: "It does NOT auto-include message field values … This is by design — message contents may contain PII or large payloads."

### Why the existing guard does not cover it

The tap is careful about the tracer itself: it snapshots `previousTracer` and restores it exactly (SpanTap.ts:70, 106-107), and `TeeTracer` keeps the original exporter working — so the one piece of state the author explicitly reasoned about is handled. `TracingExtension.disable()` does reset both flags, and it *is* reached when DevTools was the first thing to enable tracing, which is why the no-tracer case behaves correctly and is presumably why tests pass. `ActorCell` also excludes DevTools' own internal actors from spans (ActorCell.ts:936). What is missing is a symmetric restore for the two flags on the `enable(previousTracer)` branch: nothing in `SpanTap`, `TracingExtension`, or `DevToolsServer.stop()` snapshots `isRecordingRootSpans()` / `isCapturingMessagePayloads()` before overwriting them, so there is no state to restore even if the branch wanted to.

### Suggested fix

In `SpanTap.install()`, snapshot the two flags (`tracing.isRecordingRootSpans()`, `tracing.isCapturingMessagePayloads()`) alongside `previousTracer`, and restore both in `uninstall()` on every branch. Additionally, gate payload capture behind an explicit DevTools option that defaults to *off* (`DevToolsOptions.withCapturePayloads(true)`), because forwarding whole message bodies to a third-party exporter is a materially different act from opening a span — and log a one-line warning when the framework starts attaching `actor.message.payload` to spans that a non-DevTools tracer will export.

### Verification status

Found in the second, independent whole-framework security re-audit of 2026-08-02 (`v0.12.0`) — a fresh pass run without reference to the first wave's findings, then triaged against the existing tracker and adjudicated by verifiers instructed to refute it.

<details><summary>Verifier note</summary>

CONFIRMED — I reproduced the asymmetry by reading, not by trusting the evidence block.

Citation accuracy: exact. `src/devtools/taps/SpanTap.ts:86-87` sets `tracing.recordRootSpans(true)` and `tracing.captureMessagePayloads(true)` unconditionally at the end of `install()`, after the `previousTracer === null ? RecordingTracer : TeeTracer` branch (SpanTap.ts:70-81). `uninstall()` at SpanTap.ts:106-107 is `if (this.previousTracer === null) tracing.disable(); else tracing.enable(this.previousTracer);` and touches neither flag. `src/tracing/TracingExtension.ts:24-27` shows `enable()` assigning only `this.tracer`; `disable()` at TracingExtension.ts:77-82 is the sole place that clears `_traceRootSpans` / `_traceMessagePayloads`. So the reset is reachable only on the branch where DevTools was the first thing to enable tracing.

The guard the finder says is missing: I looked for it in four places and it is genuinely absent. (a) `SpanTap` never reads `isRecordingRootSpans()` / `isCapturingMessagePayloads()`, so there is no snapshot to restore — only `previousTracer` is captured (SpanTap.ts:58, 70). (b) `DevToolsServer.stop()` (DevToolsServer.ts:325-344) only loops `tap.uninstall()`; it does not touch the tracing extension. (c) A repo-wide grep for `captureMessagePayloads|recordRootSpans` in `src/` returns exactly two call sites, both SpanTap.ts:86-87 — nothing anywhere else turns them off. (d) `src/ActorSystem.ts:89,100` declares both fields with a `false` initializer and no other writer.

Test coverage gap confirms the finder's "presumably why tests pass" hypothesis. `tests/unit/devtools/SpanTap.test.ts:372-389` does assert `isCapturingMessagePayloads() === false` after `uninstall()` — but it builds a fresh system with no tracer (line 373), so it exercises only the `disable()` branch. The sibling test that does cover the pre-existing-tracer branch, `puts the original tracer back on uninstall` (lines 76-84), asserts on `tracerOf(system)` alone and never looks at the flags. The uncovered branch is precisely the buggy one.

Reachability: holds. `DevToolsServer.start()` constructs the `SpanTap` under `isPanelEnabled('tracing')` (DevToolsServer.ts:235-244), and `isPanelEnabled` (DevToolsServer.ts:161-163) is true unless the operator sets `panels.tracing === false` — i.e. on by default. `bind()` returns `detach: () => this.stop()` (DevToolsServer.ts:320), so a normal attach/detach reaches `uninstall()`. The consequence at `src/internal/ActorCell.ts:939-957` is as described: with `_traceRootSpans` true every non-`_internal` cell opens an `actor.receive` span even with no inbound trace context, and with `_traceMessagePayloads` true it attaches `actor.message.payload` = `describeMessagePayload(message)` — `JSON.stringify` bounded to depth 6 and 2000 chars per `src/internal/Instrumentation.ts:15-17,53-63`. The restored tracer is the app's real exporter, so the payloads ship out. Nothing upstream stops it: no cap, no re-check, no log line.

What I discounted. The precondition needs DevTools attached to a tracing-enabled system, and `docs/src/content/docs/observability/devtools/overview.mdx:77` explicitly advises attaching "behind a flag so production never runs it" — so the production shape in the exploit is a discouraged configuration, though the docs also cover exposing it behind auth (overview.mdx:170-176), and the same leak lands on any long-lived staging system that exports traces. No attacker is involved and the sink is the operator's own backend. Against that: the exposure is silent (no log, no metric, only `isCapturingMessagePayloads()` reveals it), persists until process restart, survives every subsequent attach/detach (the second attach also takes the `else` branch), and defeats the framework's own documented privacy default. MEDIUM is the right band — above LOW because it is silent and unbounded in time, below HIGH because it needs an operator action on a discouraged configuration and there is no external adversary.

One note on the suggested fix: restoring saved flags in `uninstall()` is safe on both branches — `recordRootSpans(false)` cannot hit the `needs a tracer` throw at TracingExtension.ts:50-54, which only fires for `enabled === true`.

**Correction applied:** Two framing points corrected. (1) The finding calls the `actor-tracing.mdx` caution ("It does NOT auto-include message field values … message contents may contain PII") a "now false" premise. That caution describes the framework default with no DevTools involved and remains true there; it becomes false only for a process that has completed one DevTools attach/detach cycle. The genuinely contradicted sentence is `docs/src/content/docs/observability/devtools/tracing.mdx:70-74` ("it does so for as long as DevTools is attached"), which is false on the pre-existing-tracer branch. (2) "third-party exporter" overstates the trust boundary: the destination is the operator's own configured tracer, not an attacker-controlled sink — the exposure is a scope expansion of message bodies into whatever APM the app already exports to (frequently a SaaS with different retention and access controls than the app), not exfiltration. Severity MEDIUM is correct as filed; no adjustment.

</details>

**Comment by pathosDev** [comment]
Independent verification of the landed fix (0d524382, on develop via #421429b0). The reported defect is genuinely fixed: reproducing the exploit walkthrough through the real `DevTools.attach(...)` → `detach()` path with a `RecordingTracer` standing in for the OTel exporter gives `baseline false/false | attached true/true | detached false/false`, zero spans exported after the detach and zero carrying `actor.message.payload`; the same probe against the parent commit leaks the planted card number on a span the application's exporter receives. Reverting `src/devtools/taps/SpanTap.ts` while keeping the new tests turns four of them red, and neutralising each added mechanism one at a time turns nine of ten mutants red. Three findings worth recording here.

1. The comment at `src/devtools/taps/SpanTap.ts:136-139` — "It cannot throw here — `false` never does, and `true` is only reachable when a tracer was installed at `install()` time" — is not true, and the commit body repeats it. `_traceRootSpans === true` with `isEnabled() === false` is reachable from the public API: `TracingExtension.enable(NOOP_TRACER)` (exported at `src/tracing/index.ts:17`) swaps the tracer for the no-op and, unlike `disable()`, clears neither flag. Then `enable(RecordingTracer) → recordRootSpans(true) → enable(NOOP_TRACER) → tap.install() → tap.uninstall()` throws `recordRootSpans(true) needs a tracer`. This is new — before this commit `uninstall()` never called `recordRootSpans` and could not throw. It matters because `DevToolsServer.stop()` sets `this.stopped = true` at `DevToolsServer.ts:433` *before* the tap loop at `:434`, so the throw skips `taps.clear()`, the ExplainMethods/agent/federation/sampler teardown, `binding.unbind()` and stopping the hub actor, and `stopped` blocks a retry. End to end: `detach()` rejects, `isAttached()` reports false, and `GET /` on the DevTools port still returns HTTP 200 serving the UI — an unauthenticated debugger surface surviving a detach the operator believes succeeded. A guard on `this.previousRootSpans && tracing.isEnabled()`, or restoring `captureMessagePayloads` first so a throw cannot skip it, plus a test, would close it.

2. The restore is not re-entrant, and one ordering is strictly worse than before the fix. With two `SpanTap`s on one system detached out of order — A installs (no tracer), B installs (snapshots A's tracer and both flags as `true`), A uninstalls (`disable()` clears everything), B uninstalls — B's restore re-enables A's dead `RecordingTracer` and sets both switches back to `true` with no DevTools attached: exactly the #714 end state. The parent commit leaves them `false`. Two taps are reachable: `DevToolsExtension.attach()` guards on `this.binding` and `mount()` guards on `this.server`, so `DevTools.mount(system, …)` followed by `DevTools.attach(system, …)` builds a second `DevToolsServer` with a second `SpanTap` (and orphans the first). Nothing in the suite installs two taps.

3. Two smaller ones. `docs/.../observability/tracing/actor-tracing.mdx:190-198` (DE `:193-201`) still states "It does NOT auto-include message field values … message contents may contain PII" with no cross-reference to the attach that suspends it — the new caution lives only on the DevTools tracing page, which is not where a reader on the OTel-adapter path lands. And the deferred third suggested-fix item (the `DevToolsOptions` payload gate) exists only as a sentence in the commit body: the promised follow-up issue was never opened, and nothing at `SpanTap.ts:107-108` records the decision.

Everything else checks out: CHANGELOG entry present in `[Unreleased] → Security` with prose that matches the code; both language mirrors of the tracing panel page updated with equivalent structure (251/251 pages, `Aside` imported in both); no new HOCON key; `Refs` rather than `Closes`, which is right while item 2 of the suggested fix is deferred.
